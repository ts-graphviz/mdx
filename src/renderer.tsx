/* eslint-disable class-methods-use-this */
import React, { createElement, FC } from 'react';
import { renderToString } from 'react-dom/server';
import MDX from '@mdx-js/runtime';
import { MDXProvider } from '@mdx-js/react';
import { Graph, Digraph, Subgraph, Node, Edge, ClusterPortal, DOT } from '@ts-graphviz/react';
import { Graphviz } from './components/Graphviz';
import { Components, Plugin } from './types';
import { GraphvizMDXProviderComponentsContext } from './contexts/GraphvizMDXProviderComponents';

export class Renderer {
  private mdx: Components = { Graphviz };

  private graphviz: Components = {
    Graph,
    Digraph,
    Subgraph,
    Node,
    Edge,
    ClusterPortal,
  };

  private remark: any[] = [];

  private rehype: any[] = [];

  public use(plugin: Plugin): this {
    Object.assign(this.mdx, plugin.mdx);
    Object.assign(this.graphviz, plugin.graphviz);
    this.remark.push(plugin.remark ?? []);
    this.rehype.push(plugin.rehype ?? []);
    return this;
  }

  public renderToHTML(mdx: string): string {
    return renderToString(
      <GraphvizMDXProviderComponentsContext.Provider value={this.graphviz}>
        <MDXProvider components={this.mdx}>
          <MDX remarkPlugins={this.remark} rehypePlugins={this.rehype}>
            {mdx}
          </MDX>
        </MDXProvider>
      </GraphvizMDXProviderComponentsContext.Provider>,
    );
  }
}
