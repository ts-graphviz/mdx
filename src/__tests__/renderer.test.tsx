import React, { FC } from 'react';
import { Node } from '@ts-graphviz/react';
import dedent from 'ts-dedent';
import { Renderer } from '../renderer';

jest.mock('@ts-graphviz/node', () => ({
  renderDot() {
    // eslint-disable-next-line no-buffer-constructor
    return new Buffer('test');
  },
}));

describe('Renderer', () => {
  let renderer: Renderer;
  beforeEach(() => {
    renderer = new Renderer();
  });
  describe('renderToHTML', () => {
    test('Digraph', () => {
      const mdx = dedent`
      # Test

      This is test MDX.

      <Graphviz>
        <Digraph>
          <Node id="node1" />
          <Node id="node2" />
          <Node id="node3" />
          <Edge targets={['node1', 'node2']} />
          <Edge targets={['node1', 'node3']} />
        </Digraph>
      </Graphviz>
      `;
      const html = renderer.renderToHTML(mdx);
      expect(html).toMatchSnapshot();
    });

    test('Graph', () => {
      const mdx = dedent`
      # Test

      This is test MDX.

      <Graphviz>
        <Graph>
          <Subgraph id="cluster_test">
            <Node id="node1" />
            <Node id="node2" />
          </Subgraph>
          <Node id="node3" />
          <Edge targets={['node1', 'node2']} />
          <Edge targets={['node1', 'node3']} />
        </Graph>
      </Graphviz>
      `;
      const html = renderer.renderToHTML(mdx);
      expect(html).toMatchSnapshot();
    });
  });

  test('register custom component', () => {
    // eslint-disable-next-line react/prop-types
    const TestNode: FC<{ id: string }> = ({ id }) => {
      return <Node id={id} fontcolor="red" />;
    };
    renderer.use({
      graphviz: { TestNode },
    });

    const mdx = dedent`
    # Test

    This is test MDX.

    <Graphviz>
      <Digraph>
        <TestNode id="node1" />
        <TestNode id="node2" />
        <TestNode id="node3" />
        <Edge targets={['node1', 'node2']} />
        <Edge targets={['node1', 'node3']} />
      </Digraph>
    </Graphviz>
    `;
    const html = renderer.renderToHTML(mdx);
    expect(html).toMatchSnapshot();
  });
});
