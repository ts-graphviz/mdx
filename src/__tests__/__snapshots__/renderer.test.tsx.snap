// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renderer register custom component 1`] = `
<h1>
  Test
</h1>
<p>
  This is test MDX.
</p>
<img src="data:image/png;base64,dGVzdA==">
`;

exports[`Renderer renderToHTML Digraph 1`] = `
<h1>
  Test
</h1>
<p>
  This is test MDX.
</p>
<img src="data:image/png;base64,dGVzdA==">
`;

exports[`Renderer renderToHTML Graph 1`] = `
<h1>
  Test
</h1>
<p>
  This is test MDX.
</p>
<img src="data:image/png;base64,dGVzdA==">
`;
